src/de.hpi.swa.graal.squeak/src/de/hpi/swa/graal/squeak/model/CompiledCodeObject.java

```diff
@@ -215,6 +215,23 @@ public final FrameSlot getStackSlot(final int i) {
         return stackSlots[i];
     }
 
+    public final FrameSlot getStackSlot(final int i, final Object value) {
+        assert 0 <= i && i < stackSlots.length : "Bad stack access";
+        if (stackSlots[i] == null) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            if (value instanceof Boolean) {
+                stackSlots[i] = frameDescriptor.addFrameSlot(i + 1, FrameSlotKind.Boolean);
+            } else if (value instanceof Long) {
+                stackSlots[i] = frameDescriptor.addFrameSlot(i + 1, FrameSlotKind.Long);
+            } else if (value instanceof Double) {
+                stackSlots[i] = frameDescriptor.addFrameSlot(i + 1, FrameSlotKind.Double);
+            } else {
+                stackSlots[i] = frameDescriptor.addFrameSlot(i + 1, FrameSlotKind.Object);
+            }
+        }
+        return stackSlots[i];
+    }
+
```

Comment:
I'm afraid we won't gain much performance from this as adding stack slots is a slow path operation anyway, so it doesn't really matter if we set the type later in the same AST or right after creation. The stack slots logic is already quite complex, so I rather not add additional logic we have to keep in tact.